test/images/agnhost/guestbook/guestbook.go

```diff
@@ -0,0 +1,278 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package guestbook
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+)
+
+// CmdGuestbook is used by agnhost Cobra.
+var CmdGuestbook = &cobra.Command{
+	Use:   "guestbook",
+	Short: "Creates a HTTP server with various endpoints representing a guestbook app",
+	Long: `Starts a HTTP server on the given --http-port (default: 80), serving various endpoints representing a guestbook app. The endpoints and their purpose are:
+
+- /register: A guestbook slave will subscribe to a master, to its given --slaveof endpoint. The master will then push any updates it receives to its registered slaves through the --backend-port.
+- /get: Returns '{"data": value}', where the value is the stored value for the given key if non-empty, or the entire store.
+- /set: Will set the given key-value pair in its own store and propagate it to its slaves, if any. Will return '{"data": "Updated"}' to the caller on success.
+- /guestbook: Will proxy the request to agnhost-master if the given cmd is 'set', or agnhost-slave if the given cmd is 'get'.`,
+	Args: cobra.MaximumNArgs(0),
+	Run:  main,
+}
+
+var (
+	httpPort    string
+	backendPort string
+	slaveOf     string
+	slaves      []string
+	store       map[string]interface{}
+)
+
+const (
+	timeout = time.Duration(15) * time.Second
+	sleep   = time.Duration(1) * time.Second
+)
+
+func init() {
+	CmdGuestbook.Flags().StringVar(&httpPort, "http-port", "80", "HTTP Listen Port")
+	CmdGuestbook.Flags().StringVar(&backendPort, "backend-port", "6379", "Backend's HTTP Listen Port")
+	CmdGuestbook.Flags().StringVar(&slaveOf, "slaveof", "", "The host's name to register to")
+	store = make(map[string]interface{})
+}
+
+func main(cmd *cobra.Command, args []string) {
+	go registerNode(slaveOf, backendPort)
+	startHTTPServer(httpPort)
+}
+
+func registerNode(registerTo, port string) {
+	if registerTo == "" {
+		return
+	}
+
+	hostPort := net.JoinHostPort(registerTo, backendPort)
+	_, err := net.ResolveTCPAddr("tcp", hostPort)
+	if err != nil {
+		log.Fatalf("--slaveof param and/or --backend-port param are invalid. %v", err)
+		return
+	}
+
+	start := time.Now()
+	for time.Since(start) < timeout {
+		response, err := dialHTTP("register", hostPort)
+		if err != nil {
+			log.Printf("encountered error while registering to master: %v. Retrying in 1 second.", err)
+			time.Sleep(sleep)
+			continue
+		}
+
+		responseJSON := make(map[string]interface{})
+		err = json.Unmarshal([]byte(response), &responseJSON)
+		if err != nil {
+			log.Fatalf("Error while unmarshaling master's response: %v", err)
+		}
+
+		store = responseJSON["data"].(map[string]interface{})
```

Comment:
Should we check whether the cast doesn't have error ?

Reply:
Done